### PR TITLE
Feature amesos2 klu2 and basker complex optimization

### PR DESCRIPTION
**Pull request by ndellingwood**
Enable single process optimizations in KLU2 and Basker when Complex is enabled.
Retrieving the sparse matrix values pointer may yield a `Kokkos::complex<T>*` which cannot be converted to `std::complex<T>*`; `reinterpret_cast` used to enable this conversion.

### CODE REVIEW

**Comment by ndellingwood** [comment]
@hkthorn This change should regain performance when complex is enabled for the use case you reported.

File: packages/amesos2/src/Amesos2_Basker_def.hpp [hunk 1]
```diff
@@ -179,8 +183,16 @@ Basker<Matrix,Vector>::symbolicFactorization_impl()
...
             std::runtime_error, "Amesos2 Runtime Error: sp_colind returned null ");
+#ifndef HAVE_TEUCHOS_COMPLEX
         auto sp_values = this->matrixA_->returnValues();
-          TEUCHOS_TEST_FOR_EXCEPTION(sp_values == nullptr,
+#else
+    // NDE: 09/25/2017
+    // Cannot convert Kokkos::complex<T>* to std::complex<T>*; in this case, use reinterpret_cast
+        using complex_type = typename Util::getStdCplxType< magnitude_type, typename matrix_adapter_type::spmtx_vals_t >::type;
+        complex_type * sp_values = nullptr;
+        sp_values = reinterpret_cast< complex_type * > ( this->matrixA_->returnValues() );
```
Comment (ndellingwood) [inline packages/amesos2/src/Amesos2_Basker_def.hpp]: @mhoemmen if you have a moment could you comment on my use of r`einterpret_cast` here, needed to convert `Kokkos::complex<T>*` to `std::complex<T>*`, specifically is there a better way I should be going about this or any possible bad situations you may foresee?

**Comment by ndellingwood** [comment]
Updated branch to incorporate recent changes to split ShyLU into subpackages

**Comment by ndellingwood** [comment]
Some issues running the checkin script due to a conflict. 
I first merged this branch into the develop branch using `git merge feature-amesos2complexoptim` while on the develop branch. No issues. 
I tested with `./checkin-test-sems.sh --pull` which initially failed due to a conflict. 
I fixed the conflict, ran the commands `git add file_with_conflict` and `git rebase --continue`
Rerunning `./checkin-test-sems.sh --pull` now works, but I'm not sure if this was the right thing to do...
@bartlettroscoe or @mhoemmen - Is my procedure above the right way to handle a conflicts (I don't want to run checkin-test-sems and push something in that will cause conflicts or issues if the test passes and pushes)? If this is not correct, what is the proper way to handle conflicts before resuming checkin-test-sems?

**Comment by bartlettroscoe** [comment]
If you are merging branch with a lot of commits on it, you and avoid the rebase by also passing in --no-rebase.
________________________________

**Pull request merged**